File: android/app/src/main/kotlin/com/example/seabattle/Accelerometer.kt
package com.example.seabattle

import android.content.Context
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager

class Accelerometer(
    private val context: Context
) : AccelerometerApi, SensorEventListener {

    private val sensorManager = context.getSystemService(Context.SENSOR_SERVICE) as SensorManager
    private var accelerometerSensor: Sensor? = null
    private var callback: AccelerometerDataCallback? = null
    private var isRunning = false

    fun setCallback(cb: AccelerometerDataCallback) {
        callback = cb
    }

    // Реализация Pigeon API
    override fun startAccelerometer(callback: (Result<Boolean>) -> Unit) {
        if (isRunning) {
            callback(Result.success(true))
            return
        }

        accelerometerSensor = sensorManager.getDefaultSensor(Sensor.TYPE_ACCELEROMETER)
        if (accelerometerSensor == null) {
            callback(Result.success(false))
            return
        }

        sensorManager.registerListener(this, accelerometerSensor, SensorManager.SENSOR_DELAY_GAME)
        isRunning = true
        callback(Result.success(true))
    }

    override fun stopAccelerometer(callback: (Result<Boolean>) -> Unit) {
        if (!isRunning) {
            callback(Result.success(true))
            return
        }

        sensorManager.unregisterListener(this)
        isRunning = false
        callback(Result.success(true))
    }

    override fun onSensorChanged(event: SensorEvent?) {
        if (event == null || callback == null) return

        val data = AccelerometerData(
            x = event.values[0].toDouble(),
            y = event.values[1].toDouble(),
            z = event.values[2].toDouble()
        )

        callback?.onAccelerometerDataReceived(data) {
            // здесь можно игнорировать результат
        }
    }

    override fun onAccuracyChanged(sensor: Sensor?, accuracy: Int) {}
}

File: ios/Runner/Accelerometer.swift
import Foundation
import CoreMotion
import Flutter

class Accelerometer: NSObject, AccelerometerApi {

    private let motionManager = CMMotionManager()
    private var callback: AccelerometerDataCallback? = nil
    private var isRunning = false

    func setCallback(callback: AccelerometerDataCallback) {
        self.callback = callback
    }

    func startAccelerometer(completion: @escaping (Result<Bool, Error>) -> Void) {
        if isRunning {
            completion(.success(true))
            return
        }

        guard motionManager.isAccelerometerAvailable else {
            completion(.success(false))
            return
        }

        motionManager.accelerometerUpdateInterval = 1.0 / 60.0

        motionManager.startAccelerometerUpdates(to: .main) { data, error in
            guard let data = data else { return }

            let acc = AccelerometerData(
                x: -data.acceleration.x,
                y: -data.acceleration.y,
                z: data.acceleration.z
            )

            self.callback?.onAccelerometerDataReceived(data: acc) { _ in
                // результат игнорируем, как на Android
            }
        }

        isRunning = true
        completion(.success(true))
    }

    func stopAccelerometer(completion: @escaping (Result<Bool, Error>) -> Void) {
        if !isRunning {
            completion(.success(true))
            return
        }

        motionManager.stopAccelerometerUpdates()
        isRunning = false
        completion(.success(true))
    }
}
